### PR TITLE
fix(code-snippet): add light styles for all variations and update storybook knobs

### LINKED ISSUE #4341: CodeSnippet - `light` prop only works for `inline` variation
## What package(s) are you using?

- [x] `carbon-components`
- [x] `carbon-components-react`

## Detailed description

In the React package, the `light` prop is available to all variations of the `CodeSnippet`, even though the comment mentions typically applying it to the `inline` variation: https://github.com/carbon-design-system/carbon/blob/27fa8eccf4442da209c19662ed211d9a3fc78331/packages/react/src/components/CodeSnippet/CodeSnippet.js#L76-L81

However, the `light` prop & applied `bx--snippet--light` class don't change any `CodeSnippet` styles for single or multi-line variations.

It turns out that there are only `light` styles for the `inline` variation:
https://github.com/carbon-design-system/carbon/blob/27fa8eccf4442da209c19662ed211d9a3fc78331/packages/components/src/components/code-snippet/_code-snippet.scss#L62-L68

I realize that the expected use-case is for only the `inline` variation to apply the `light` prop, but in our applications we need to be able to use the `light` prop for multi-line code snippets. Our application developers see a prop that unfortunately doesn't do anything when they use it.

### PR DESCRIPTION
**Pull request by jendowns**
Closes #4341 

Proposal to add light styles to all variations of the `CodeSnippet`, since the `light` prop is available to use for the component but currently only applies styles for the `inline` variation.

#### Changelog

**Changed**

- update `.bx--snippet--light` styles to apply to all variations of the `CodeSnippet` (and also update button hover colors to accurately reflect these changes) -- **NOTE:** I realize the original `inline` + `light` variation used `$field-01` props, however the `$ui-*` tokens I picked out are the same and are good foils for the non-`light` `$ui-*` tokens used.
- update story to include `light` prop knobs for all variations of `CodeSnippet`

#### Testing / Reviewing

Check out the `carbon-components-react` Netlify deployment for this PR & use the knobs to toggle the `light` state for each `CodeSnippet` variation.

### CODE REVIEW

File: packages/components/src/components/code-snippet/_code-snippet.scss [hunk 1]
```diff
@@ -59,14 +59,6 @@
     padding: 0 $spacing-xs;
   }
 
-  .#{$prefix}--snippet--inline.#{$prefix}--snippet--light {
-    background-color: $field-02;
-
-    &:hover {
-      background-color: rgba($interactive-02, 0.1);
```
Comment (jendowns) [inline packages/components/src/components/code-snippet/_code-snippet.scss:66]: **Please note:** this style rule isn't valid & currently the `inline` + `light` variation of the `CodeSnippet` doesn't have a background-color on hover as a result: http://react.carbondesignsystem.com/?path=/story/codesnippet--inline

Screen Shot 2019-10-15 at 2 12 36 PM

**Review by asudoh — APPROVED** [review]
LGTM 👍 - Thanks @jendowns!

**Review by emyarod — APPROVED** [review]
looks good to me, pending design review

File: packages/components/src/components/code-snippet/_code-snippet.scss (hunk 1 above)
Comment (emyarod) [reply packages/components/src/components/code-snippet/_code-snippet.scss:66]: I believe this is valid Sass but these rules will need tweaking for outputting custom properties

Comment (jendowns) [reply packages/components/src/components/code-snippet/_code-snippet.scss:66]: When you say it's valid, are you referring to using a hex value inside the color rgba function? Like: `rgba(#000000, 0.5)`? 

I don't believe that's correct. I opened an issue yesterday about this that goes into more detail (and mentions the `Loading` overlay background-color, where similar `rgba` usage isn't working either: https://github.com/carbon-design-system/carbon/issues/4345)

A quick way to test this is to open the inspector in this page, for example, and try adding `background-color: rgba(#00000, 0.5)` -- in Chrome, that doesn't work & the inspector marks it as invalid.

Screen Shot 2019-10-16 at 9 21 59 AM


But you can change that to `background-color: rgba(0, 0, 0, 0.5)` and it works.

Comment (jendowns) [reply packages/components/src/components/code-snippet/_code-snippet.scss:66]: Worth noting btw that I'm not sure it's the token usage inside the `rgba` that is causing issues -- it's the fact that the tokens reference hex values, not rgb values, that is the problem.

Comment (emyarod) [reply packages/components/src/components/code-snippet/_code-snippet.scss:66]: it is [valid Sass](https://sass-lang.com/documentation/modules#rgb) and not valid CSS. previously the output would be converted from hex to rgb values with the Sass `rgba` function (Sass => CSS). after the recent theming work we are emitting custom properties with fallback values (Sass => custom properties => CSS) but the fallbacks are still referencing Sass functions

**Review by dakahn — CHANGES_REQUESTED** [review]
[WCAG 1.4.11](https://www.w3.org/WAI/WCAG21/Understanding/non-text-contrast.html) requires a 3:1 contrast ratio for non-text interface elements and their adjacent colors. Common ways of meeting this requirement are 3:1 ratio contrast borders or improving the contrast between the interface element and it's adjacent colors (page background, parent background etc). 

Currently our code snippet does not meet these requirements. While most code-snippets are not considered "interface components" ours _are_ since they are interactable. 

This situation is made worse in lighter themes by allowing an even lower contrast variant background color for these interface components.

**Comment by jendowns** [comment]
@dakahn could you describe which parts of the `light` variation of the `CodeSnippet` you are referring to?

**Comment by dakahn** [comment]
@jendowns The interface element's background color must meet a 3:1 contrast ratio against adjacent colors.

**Comment by jendowns** [comment]
@dakahn 🤔 So do you mean that a "light" `CodeSnippet` technically can't really exist, because that background would be too close to the `$ui-background`?

**Comment by dakahn** [comment]
@dakahn precisely. Since it's interactable it must meet 1.4.11 contrast requirements (3:1) for non-text interface elements. Worth noting -- it currently doesn't meet this requirement today.

**Comment by jendowns** [comment]
@dakahn Are we definitely sure we should consider all variations of the `CodeSnippet` interactive?

I can definitely see it for the `inline` one because you click it to copy etc.
But what about the `single` and `multi-line` versions? Technically they have interactive buttons, but the body of the components (where code is) isn't interactive? 🤔

**Comment by jendowns** [comment]
Also something else to consider... in our case, we are using the `light` prop to distinguish the `CodeSnippet` from a `ui-01` or similar background where it would normally have been the same color as the background color.

Here's an example of a `light` + `inline` snippet with some context (in a panel):
Screen Shot 2019-10-16 at 2 30 20 PM



(so maybe what you are seeing in the storybook isn't a good in-product example of how it would be used anyway? 🤔 )

**Review by dakahn — APPROVED** [review]
Okay! If we're only using `light` on dark themes and we have docs and guidance coming to back that up. I'm cool with this. 👍

**Review by jillianhowarth — APPROVED** [review]

**Review by aledavila — APPROVED** [review]

**Comment by jendowns** [comment]
@dakahn Per our Slack convo, I added a conditional wrapper class & a `light` prop message to help people use the prop correctly. So if you look at the `CodeSnippet` variations in different themes and toggle the `light` knob, you'll see something like this:

Screen Shot 2019-10-17 at 11 21 45 AM
Screen Shot 2019-10-17 at 11 21 31 AM

**Comment by dakahn** [comment]
@joshblack How do you feel about this in-Storybook warning. I thought this would be cool as a stop-gap until we figure out how we're doing guidance in this thing.

**Comment by joshblack** [comment]
I'm down if it helps 👍 what do you think @jillianhowarth? Is this okay from the design perspective to help educate devs who might mistakenly use this prop?

File: packages/react/src/components/CodeSnippet/CodeSnippet-story.js [hunk 2]
```diff
@@ -48,12 +50,23 @@ const props = {
   }),
 };
 
+const lightPropMessage = (
+  <small style={{ display: 'block', paddingBottom: '1rem' }}>
+    {'Make sure to use a different background color for the '}
```
Comment (jillianhowarth) [inline packages/react/src/components/CodeSnippet/CodeSnippet-story.js]: This copy is unclear - we should clarify that they should be using the prop with certain themes

Comment (jendowns) [reply packages/react/src/components/CodeSnippet/CodeSnippet-story.js]: Hey @jillianhowarth the prop is currently available regardless of theme. I'm happy to change the copy though -- what should it be?

Comment (aagonzales) [reply packages/react/src/components/CodeSnippet/CodeSnippet-story.js]: It's safe to use with all themes actually. It's working correctly in the White and Gray 10 themes as well. 

If we want to get more specific/technical we could say something like: 
- The snippet container should never be the same color as the page background. 
- Do not use the `light` variant on `$ui-background` or `$ui-02`.

**Review by aagonzales — COMMENTED** [review]
The hover of the light snippet should be using `hover-ui` as well so it doesn't blind in.

**Comment by jendowns** [comment]
Thanks @aagonzales & @jillianhowarth!

I have updated that copy & you should see it reflected in the netlify preview soon.
But as for using `$hover-ui` for the hover color... currently it looks great for `white` and `g10` themes but it is very hard to see for the dark themes `g90` and `g100` 🤔 Seems to be because `$hover-ui` is very close to `$ui-02` (the "light" variant bg color).

Here's g90 for example (I'm hovering over the "copy" button in the upper right of the multi line snippet):

Screen Shot 2019-10-17 at 1 42 48 PM

**Comment by aagonzales** [comment]
Hmm I wonder if the new color updates made them too similar. `hover-ui` is the appropriate token so I'm wondering if we need to update that token value. Let me look at it in relation to some other things. 

Or maybe its time to add an `hover-ui-light` token specifically for the light variants 🤔 

--- 

These are the options

image

**Comment by jendowns** [comment]
@aagonzales Personally I'm leaning more toward the new token just so it doesn't affect any other components. 😅 Is that okay? If so, I could go ahead and add the new token in this PR if you happen to know what colors it should be per theme?

**Comment by aagonzales** [comment]
@jendowns actually I just talked to Josh about this, could we actually do a lighten/darken effect for the hover for these? That's the direction we're trying to go with hovers in v11 so we can get ahead of the game without adding a new token.

White and gray 10: darken by 8%
Gray 90 and 100: lighten by 8%

**Comment by jendowns** [comment]
Thank you @aagonzales! That makes sense. And for the `v9` theme, should the color be darkened by 8% as well?

**Comment by aagonzales** [comment]
Hmm for V9, the icon color darkens not the background. I think its also controlled by a different token in v9 if I'm not mistaken.

**Comment by joshblack** [comment]
@jendowns sorry about all the changes here! Let me know if you need anything else from us 👀

**Comment by joshblack** [comment]
Chatted with @jendowns, I think we'll move forward with using the value out of the theme map and lighten/darken that instead of the property while we figure out the issue around CSS Custom Properties with lighten/darken values over in https://github.com/carbon-design-system/carbon/issues/4426

File: packages/components/src/globals/scss/_theme-tokens.scss [hunk 3]
```diff
@@ -358,3 +359,41 @@ $tab-underline-disabled: 3px solid $disabled-01 !default;
...
+
+  @if $carbon--theme ==
+    $carbon--theme--g90 or
+    $carbon--theme ==
+    $carbon--theme--g100
+  {
+    $value: lighten($value, 8%);
+  }
+
+  @if map-get($feature-flags, enable-css-custom-properties) == true {
```
Comment (joshblack) [inline packages/components/src/globals/scss/_theme-tokens.scss]:
```suggestion
  @if feature-flags-enabled('enable-css-custom-properties') {
```

File: packages/components/src/globals/scss/_theme-tokens.scss [hunk 4]
```diff
@@ -358,3 +359,41 @@ $tab-underline-disabled: 3px solid $disabled-01 !default;
...
+    @return var(--cds-#{$token}-light, $value);
```
Comment (joshblack) [inline packages/components/src/globals/scss/_theme-tokens.scss]: Would it be worth trying to figure out a function variant of https://github.com/carbon-design-system/carbon/blob/master/packages/themes/scss/_mixins.scss#L14 ?

If not, might be able to use `$custom-property-prefix` like

```suggestion
    @return var(--#{$custom-property-prefix}-#{$token}-light, $value);
```

File: packages/components/src/components/code-snippet/_code-snippet.scss [hunk 5]
```diff
@@ -266,6 +259,26 @@
     transition: transform $transition--expansion;
   }
 
+  // Light version
+  .#{$prefix}--snippet.#{$prefix}--snippet--light {
+    &,
```
Comment (joshblack) [inline packages/components/src/components/code-snippet/_code-snippet.scss]: Would you mind un-nesting these selectors?

**Review by joshblack — CHANGES_REQUESTED** [review]

**Review by joshblack — APPROVED** [review]

**Pull request merged**